devlib/instrument/gem5stats.py

```diff
@@ -0,0 +1,152 @@
+#    Copyright 2017 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from __future__ import division
+import csv
+import re
+
+from devlib.instrument import Instrument, CONTINUOUS, MeasurementsCsv
+from devlib.exception import TargetError, HostError
+
+class Gem5StatsInstrument(Instrument):
+    '''
+    Instrument enabling power monitoring in Gem5
+
+    This instrument parses the stats file produced by Gem5 and extracts power
+    values. More information on the "stats.txt" file structure can be found 
+    here:
+    
+    http://pages.cs.wisc.edu/~david/courses/cs752/Spring2015/gem5-tutorial/part1/gem5_stats.html
+    
+    This instrument also implements Gem5's ROIs (Regions of Interest). A ROI is
+    an annotation on a time interval marked by start and end cursors. This enables
+    power analysis on a specific time window which spans over the execution of
+    a workload for example. Up to 8 ROIs can be used, with or without 
+    overlapping.
+    '''
+
+    mode = CONTINUOUS
+    dump_head='---------- Begin Simulation Statistics ----------'
+    dump_tail='---------- End Simulation Statistics   ----------'
+    
+    def __init__(self, target, channel_map, stats_file_path):
+        super(Gem5StatsInstrument, self).__init__(target)
+        
+        # Duration since last dump is indexed by 'sim_seconds' in stats.txt
+        self.add_channel('Duration', 'time', 'sim_seconds')
+        # Track power only for now
+        for f in channel_map:
+            self.add_channel(f, 'power', channel_map[f])
+        
+        self._stats_file_path = stats_file_path
+        self._roi_state = [False] * 8
+        self.dumps = []
+
+    def start(self, roi=0):
```

Comment:
Assuming that the instrument will be made more general, meaning not focusing only on power fields but also applicable on any field of the stats file, you will need to use ROIs to define start and stop of measurements anyway. You could set up this instrument to track the number of L3 cache misses while some task is running using the same start/stop/get_data usage model.

I don't think _what_ you want to measure and _when_ you want to measure it is dissociated in any of devlib's instrument so far. You always say what you're focusing on in `__init__` and when you focus on it with start/stop. That's the same thing here, ROIs are just a mean of implementing start/stop cleanly.